In [ ]:
pip install pandas

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
pip install sklearn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
pip install nltk

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [1]:
import pandas as pd 
import numpy as np 
import nltk 
from nltk.corpus import stopwords
import string

In [2]:
from google.colab import drive

In [3]:
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
dataset=pd.read_csv('/content/drive/MyDrive/MailAndPishingData.csv')
dataset

Category                                            Message
0         ham  Go until jurong point, crazy.. Available only ...
1         ham                      Ok lar... Joking wif u oni...
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...
3         ham  U dun say so early hor... U c already then say...
4         ham  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567     spam  This is the 2nd time we have tried 2 contact u...
5568      ham               Will ü b going to esplanade fr home?
5569      ham  Pity, * was in mood for that. So...any other s...
5570      ham  The guy did some bitching but I acted like i'd...
5571      ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [5]:
dataset.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

**Spam= 1
Not Spam=0**


In [6]:
dataset.shape

(5572, 2)

In [7]:
dataset.columns

Index(['Category', 'Message'], dtype='object')

In [8]:
dataset.drop_duplicates(inplace=True)

In [9]:
dataset.shape

(5157, 2)

In [10]:
dataset['spam']=dataset['Category'].apply(lambda x: 1 if x=='spam' else 0)

In [11]:
dataset

Category                                            Message  spam
0         ham  Go until jurong point, crazy.. Available only ...     0
1         ham                      Ok lar... Joking wif u oni...     0
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...     1
3         ham  U dun say so early hor... U c already then say...     0
4         ham  Nah I don't think he goes to usf, he lives aro...     0
...       ...                                                ...   ...
5567     spam  This is the 2nd time we have tried 2 contact u...     1
5568      ham               Will ü b going to esplanade fr home?     0
5569      ham  Pity, * was in mood for that. So...any other s...     0
5570      ham  The guy did some bitching but I acted like i'd...     0
5571      ham                         Rofl. Its true to its name     0

[5157 rows x 3 columns]

In [12]:
X=dataset['Message']

In [14]:
X

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                 Will ü b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: Message, Length: 5157, dtype: object

In [13]:
Y=dataset['spam']

In [15]:
Y

0       0
1       0
2       1
3       0
4       0
       ..
5567    1
5568    0
5569    0
5570    0
5571    0
Name: spam, Length: 5157, dtype: int64

In [16]:
dataset.shape

(5157, 3)

In [17]:
dataset.isnull().sum()

Category    0
Message     0
spam        0
dtype: int64

In [18]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [20]:
from sklearn.model_selection import train_test_split

In [21]:
from sklearn.feature_extraction.text import TfidfVectorizer
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,random_state=42)

In [22]:
from sklearn.feature_extraction.text import CountVectorizer
v=CountVectorizer()
cv_messages=v.fit_transform(X_train.values)
cv_messages.toarray()[0:5]

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [23]:
from sklearn.naive_bayes import MultinomialNB

In [24]:
classifier= MultinomialNB()

In [25]:
history = classifier.fit(cv_messages,Y_train)

In [26]:
#print the predicted Values
print("Predicted Values")
print(history.predict(cv_messages))

#print the Actual Values
print("Actual Values")
print(Y_train.values)

Predicted Values
[0 0 0 ... 0 0 0]
Actual Values
[0 0 0 ... 0 0 0]


In [27]:
from sklearn.metrics import classification_report , confusion_matrix , accuracy_score

In [28]:
pred=history.predict(cv_messages)
print(classification_report(Y_train,pred))

              precision    recall  f1-score   support

           0       0.99      1.00      1.00      3620
           1       0.98      0.96      0.97       505

    accuracy                           0.99      4125
   macro avg       0.99      0.98      0.98      4125
weighted avg       0.99      0.99      0.99      4125



In [29]:
print('Confusion Matrix : \n' , confusion_matrix(Y_train,pred))

Confusion Matrix : 
 [[3612    8]
 [  19  486]]


In [30]:
print('Accuracy' ,accuracy_score(Y_train,pred))

Accuracy 0.9934545454545455


In [31]:
test=v.transform(X_test)


In [32]:
test_data=["SMS. ac Sptv: The New Jersey Devils and the Detroit Red Wings play Ice Hockey. Correct or Incorrect? End? Reply END SPTV"]


In [33]:
input_data=v.transform(test_data)


In [34]:
pred_test1=history.predict(test)

In [35]:
print('Accuracy' ,accuracy_score(Y_test,pred_test1))

Accuracy 0.9883720930232558


In [36]:
prediction=history.predict(input_data)
print(prediction)

[1]


**Accuracy of Naive Bayes is 99.31% on Training Data and 98.83% on Testing Data**

**Now Applying the LogisticRegression**

In [41]:
from sklearn.feature_extraction.text import TfidfVectorizer
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,random_state=3)

In [44]:
cv_messages=v.fit_transform(X_train.values)
cv_messages.toarray()[0:5]

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [46]:
print(cv_messages)

  (0, 2461)	1
  (0, 1287)	1
  (0, 6819)	1
  (0, 6913)	1
  (0, 3405)	1
  (1, 950)	1
  (1, 2343)	1
  (1, 7651)	1
  (1, 3332)	1
  (1, 982)	1
  (1, 4914)	1
  (1, 6774)	1
  (1, 1603)	1
  (1, 6694)	1
  (1, 4385)	1
  (1, 3490)	1
  (1, 6898)	1
  (1, 6045)	1
  (1, 1644)	1
  (2, 6898)	1
  (2, 5969)	1
  (2, 4164)	1
  (2, 7157)	1
  (2, 4215)	1
  (2, 4665)	1
  :	:
  (4122, 6898)	1
  (4122, 3612)	1
  (4122, 6790)	1
  (4122, 3721)	1
  (4122, 4231)	1
  (4122, 3235)	1
  (4122, 3941)	1
  (4122, 6317)	1
  (4122, 6796)	1
  (4122, 4476)	1
  (4122, 6015)	1
  (4123, 6774)	1
  (4123, 1248)	1
  (4123, 6778)	1
  (4123, 5367)	1
  (4123, 4608)	1
  (4123, 7624)	1
  (4123, 1305)	1
  (4123, 1647)	1
  (4123, 5899)	1
  (4124, 3332)	1
  (4124, 6778)	1
  (4124, 6810)	1
  (4124, 4835)	1
  (4124, 7575)	1


In [47]:
from sklearn.linear_model import LogisticRegression

model=LogisticRegression()
model.fit(cv_messages,Y_train)

LogisticRegression()

In [48]:
pred_train1=model.predict(cv_messages)

In [49]:
print('Accuracy' ,accuracy_score(Y_train,pred_train1))

Accuracy 0.9980606060606061


In [50]:
test=v.transform(X_test)
pred_test1=model.predict(test)

In [51]:
print(Y_test)

464     0
2543    0
2951    0
4064    0
833     1
       ..
1402    0
1234    0
4442    0
985     0
1601    0
Name: spam, Length: 1032, dtype: int64


In [52]:
print('Accuracy' ,accuracy_score(Y_test,pred_test1))

Accuracy 0.9767441860465116


**Accuracy of Logistic Regression is 99.80% on Training Data and 97.67% on Testing Data**

In [53]:
test_data=["-PLS STOP bootydelious (32/F) is inviting you to be her friend. Reply YES-434 or NO-434 See her: www.SMS.ac/u/bootydelious STOP? Send STOP FRND to 62468"]

In [54]:
input_data=v.transform(test_data)

In [55]:
prediction=model.predict(input_data)
print(prediction)

[1]


In [56]:
import pickle

In [57]:
from sklearn.pipeline import Pipeline
clf=Pipeline([
              ('vectorizer',CountVectorizer()),
              ('nb',MultinomialNB())
]
)
clf.fit(X_train,Y_train)

Pipeline(steps=[('vectorizer', CountVectorizer()), ('nb', MultinomialNB())])

In [58]:
test_data=["Hello Boss"]
clf.predict(test_data)

array([0])

In [59]:
clf.score(X_test,Y_test)

0.9786821705426356

In [60]:
model_1 = pickle.dump(clf, open('SpamAndPishingDetection.pkl', 'wb'))

In [61]:
pickled_model = pickle.load(open('SpamAndPishingDetection.pkl', 'rb'))
pickled_model.predict(["Hello World"])

array([0])